src/README.md

```diff
@@ -0,0 +1,858 @@
+
+  return v8::Just(sum);
+}
+
+// Function that is exposed to JS:
+void SumNumbers(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  // This will crash if the first argument is not an array. Let’s assume we
+  // have performed type checking in a JavaScript wrapper function.
+  CHECK(args[0]->IsArray());
+
+  double sum;
+  if (!SumNumbers(args.GetIsolate()->GetCurrentContext(),
+                  args[0].As<v8::Array>()).To(&sum)) {
+    // Nothing to do, we can just return directly to JavaScript.
+    return;
+  }
+
+  args.GetReturnValue().Set(sum);
+}
+```
+
+#### TryCatch
+
+If there is a need to catch JavaScript exceptions in C++, V8 provides the
+`v8::TryCatch` type for doing so, which we wrap into our own
+`node::errors::TryCatchScope` in Node.js. The latter has the additional feature
+of providing the ability to shut down the program in the typical Node.js way
+(printing the exception + stack trace) if an exception is caught.
+
+<a id="libuv-handles-and-requests"></a>
+### libuv handles and requests
+
+Two central concepts when working with libuv are handles and requests.
+
+Handles are subclasses of the `uv_handle_t` “class”, and generally refer to
+long-lived objects that can emit events multiple times, such as network sockets
+or file system watchers.
+
+In Node.js, handles are often managed through a [`HandleWrap`][] subclass.
+
+Requests are one-time asynchronous function calls on the event loop, such as
+file system requests or network write operations, that either succeed or fail.
+
+In Node.js, requests are often managed through a [`ReqWrap`][] subclass.
+
+### Environment cleanup
+
+When a Node.js [`Environment`][] is destroyed, it generally needs to clean up
+any resources owned by it, e.g. memory or libuv requests/handles.
+
+<a id="cleanup-hooks"></a>
+#### Cleanup hooks
+
+Cleanup hooks are provided that run before the [`Environment`][]
+is destroyed. They can be added and removed through by using
+`env->AddCleanupHook(callback, hint);` and
+`env->RemoveCleanupHook(callback, hint);`, where callback takes a `void* hint`
+argument.
+
+Inside these cleanup hooks, new asynchronous operations *may* be started on the
+event loop, although ideally that is avoided as much as possible.
+
+Every [`BaseObject`][] has its own cleanup hook that deletes it. For
+[`ReqWrap`][] and [`HandleWrap`][] instances, cleanup of the associated libuv
+objects is performed automatically, i.e. handles are closed and requests
+are cancelled if possible.
+
+#### Closing libuv handles
+
+If a libuv handle is not managed through a [`HandleWrap`][] instance,
+it needs to be closed explicitly. Do not use `uv_close()` for that, but rather
+`env->CloseHandle()`, which works the same way but keeps track of the number
+of handles that are still closing.
+
+#### Closing libuv requests
+
+There is no way to abort libuv requests in general. If a libuv request is not
+managed through a [`ReqWrap`][] instance, the
+`env->IncreaseWaitingRequestCounter()` and
```

Comment:
I am slightly concerned direct references to these methods could get out of date. Can't we just write these down in env.h and link to the headers instead?

Reply:
> Can't we just write these down in env.h and link to the headers instead?

I’m not sure what you mean? But either way, I think keeping this doc up to date and making sure of that during reviews is something we’ll have to do either way.

Reply:
> I think keeping this doc up to date and making sure of that during reviews is something we’ll have to do either way.

I think if keeping this doc up-to-date is going to be a blocker for landing changes in `src/`, we should just mov the information about the internal classes into the headers as comments, and link to those in the README. For one if the README is too long it becomes difficult to read, also it's going to be a pain to remember updating this doc while changing the code - but slightly less so if the comments live in the headers.

Also it feels weird to me that in order to understand some of these classes, you either have to read the code, or read this doc because what's supposed to be said in comments are moved into a markdown file...

Reply:
tbh, I really like the proposed format. It has a flow, the necessary level of abstraction, example code at right places, and caveats and cautions as appropriate - something either not possible in code comments, or not easily consumable. IMO inline comments are rich with the context, but weak in connecting bigger views, and more importantly, discoverability.

Reply:
If you have items that you feel should also be documented more explicitly in a header file, I’m happy to do that, but generally I agree with @gireeshpunathil – this is not supposed to be an exhaustive guide or reference documentation, but rather paint the big picture and distinguish what’s important from what’s not.

Reply:
> and more importantly, discoverability.

I feel the opposite - I think content in the README is more difficult to discover than content in the headers, especially when the README becomes long.

> this is not supposed to be an exhaustive guide or reference documentation, but rather paint the big picture and distinguish what’s important from what’s not.

That's what I think this document should be like as well, but with details like

> They can be added and removed through by using `env->AddCleanupHook(callback, hint);` and `env->RemoveCleanupHook(callback, hint);`, where callback takes a `void* hint` argument.

and 

>  If a libuv request is not managed through a [`ReqWrap`][] instance, the `env->IncreaseWaitingRequestCounter()` and `env->DecreaseWaitingRequestCounter()` functions need to be used to keep track of the number of active libuv requests.

It does not seem to be the case. These references to specific method names and their arguments do not seem to be necessary in terms of giving the reader a bigger view of the code base.

Reply:
> > and more importantly, discoverability.
> 
> I feel the opposite - I think content in the README is more difficult to discover than content in the headers, especially when the README becomes long.

That’s somewhat solvable by adding a table of contents, if you like?

> > this is not supposed to be an exhaustive guide or reference documentation, but rather paint the big picture and distinguish what’s important from what’s not.
> 
> That's what I think this document should be like as well, but with details like
> 
> > They can be added and removed through by using `env->AddCleanupHook(callback, hint);` and `env->RemoveCleanupHook(callback, hint);`, where callback takes a `void* hint` argument.
> 
> and
> 
> > If a libuv request is not managed through a [`ReqWrap`][] instance, the `env->IncreaseWaitingRequestCounter()` and `env->DecreaseWaitingRequestCounter()` functions need to be used to keep track of the number of active libuv requests.
> 
> It does not seem to be the case. These references to specific method names and their arguments do not seem to be necessary in terms of giving the reader a bigger view of the code base.

Well … where would you document these, especially the latter one? It’s easy to add inline docs to `ReqWrap` but what do you do when you want to describe when it’s about *not* using `ReqWrap`?

In any case, I don’t feel like these give away too much detail. Judging from the current contents of the file in this PR, the number of yearly PRs that would require updating it is probably in the single digits.

Reply:
> That’s somewhat solvable by adding a table of contents, if you like?

I think that does not help in this case. To quote my edit:

> For me it's rare that I'll read a doc, top to bottom, to understand a code base and remember all these identifier names. I usually run into these names when I hack on the code and I expect to see explanations in the headers.

Now this is solvable if in the headers we point the readers to the README for more information, but that feels kind of weird.

> where would you document these, especially the latter one?

I think if the goal is to give a big picture in the README, for the latter one it can just say something like:

```
If a libuv request is not managed through a [ReqWrap][] instance, a counter is used to make sure it gets cleaned up.
```

And the method names can be documented in the `req_wrap.h`. The README can just point the readers to that header to see these details. This is a README for `src/` after all, so I'd expect it to be more of a table of contents for all these files under `src/`.